ddsketch/store.py

```diff
@@ -3,193 +3,415 @@
+        """calculate the bin index for the key, extending the range if necessary"""
+
+        if key < self.min_key:
+            self._extend_range(key)
+        elif key > self.max_key:
+            self._extend_range(key)
+
+        return key - self.offset
+
+    def _get_new_length(self, new_min_key, new_max_key):
+        desired_length = new_max_key - new_min_key + 1
+        return self.chunk_size * math.ceil((desired_length) / self.chunk_size)
+
+    def _extend_range(self, key, second_key=None):
+        """Grow the bins as necessary and call _adjust"""
+        second_key = second_key or key
+        new_min_key = min(key, second_key, self.min_key)
+        new_max_key = max(key, second_key, self.max_key)
+
+        if self.length() == 0:
+            # initialize bins
+            self.bins = [0] * self._get_new_length(new_min_key, new_max_key)
+            self.offset = new_min_key
+            self._adjust(new_min_key, new_max_key)
+
+        elif new_min_key >= self.min_key and new_max_key < self.offset + self.length():
+            # no need to change the range; just update min/max keys
+            self.min_key = new_min_key
+            self.max_key = new_max_key
+
+        else:
+            # grow the bins
+            new_length = self._get_new_length(new_min_key, new_max_key)
+            if new_length > self.length():
+                self.bins.extend([0] * (new_length - self.length()))
+            self._adjust(new_min_key, new_max_key)
+
+    def _adjust(self, new_min_key, new_max_key):
+        """Adjust the bins, the offset, the min_key, and max_key, without resizing the
+        bins, in order to try making it fit the specified range.
+        """
+        self._center_bins(new_min_key, new_max_key)
+        self.min_key = new_min_key
+        self.max_key = new_max_key
+
+    def _shift_bins(self, shift):
+        """shift the bins; this changes the offset"""
+        if shift > 0:
+            self.bins = self.bins[:-shift]
+            self.bins[:0] = [0] * shift
+        else:
+            self.bins = self.bins[abs(shift) :]
+            self.bins.extend([0] * abs(shift))
+        self.offset -= shift
+
+    def _center_bins(self, new_min_key, new_max_key):
+        """ center the bins; this changes the offset"""
+        middle_key = new_min_key + (new_max_key - new_min_key + 1) // 2
+        self._shift_bins(self.offset + self.length() // 2 - middle_key)
+
     def key_at_rank(self, rank):
-        """Return the key for the value at given rank"""
-        n = 0
-        for i, b in enumerate(self.bins):
-            n += b
-            if n >= rank:
-                return i + self.min_key
+        running_ct = 0
+        for i, bin_ct in enumerate(self.bins):
+            running_ct += bin_ct
+            if running_ct >= rank:
+                return i + self.offset
         return self.max_key
 
-    def _grow_left(self, key):
-        if self.min_key < key:
+    def reversed_key_at_rank(self, rank):
```

Comment:
Any reason to have this method, instead of using `store.key_at_rank(store.count + 1 - rank)`?

Reply:
nope!